activesupport/test/logger_test.rb

```diff
@@ -120,14 +120,73 @@ def test_buffer_multibyte
     byte_string.force_encoding("ASCII-8BIT")
     assert byte_string.include?(BYTE_STRING)
   end
-  
+
   def test_silencing_everything_but_errors
     @logger.silence do
       @logger.debug "NOT THERE"
       @logger.error "THIS IS HERE"
     end
-    
+
     assert !@output.string.include?("NOT THERE")
     assert @output.string.include?("THIS IS HERE")
   end
+
+  def test_logger_thread_safety
+    @logger.level = Logger::INFO
+
+    assert_level(Logger::INFO)
+
+    threads = (1..2).collect do |i|
+      # stagger the threads out using sleep so that they overlap during
+      # log level changes, e.g.:
+      #
+      #    Time | Thread_1        | Thread_2
+      #   ------+-----------------+-----------------
+      #    ~0.0 | 1st sleep start | 1st sleep start
+      #    ~0.1 | 1st sleep end   | <sleeping>
+      #    ~0.1 | #with_level()   | <sleeping>
+      #    ~0.1 | 2nd sleep start | <sleeping>
+      #    ~0.2 | <sleeping>      | #with_level()
+      #    ~0.2 | <sleeping>      | 2nd sleep start
+      #    ~0.3 | 2nd sleep end   | <sleeping>
+      #    ~0.4 | <dead>          | 2nd sleep end
+
+      Thread.new do
```

Comment:
I still find this test with the threads hard to understand, and I'm scared we might run into timing issues (might be overblown).

Could `stop` be of help here? That's pretty much what you're doing with `sleep`.